### DISCUSSION TITLE
Can we simulate Free-surface flows with Oceananigans.jl?

### DISCUSSION THREAD

**Discussion by kishore-nori** [CliMA/Oceananigans.jl#1292] [General]
Hi everyone! Thanks a lot for developing the package :). I wanted to know if it has methods or a framework for simulating Free-surface flows (Incompressible Newtonian NSE with gravity, in Laminar regime) using the Finite Volume method?

**Comment by francispoulin** [comment]
@kishore-nori , glad to hear you are enjoying Oceananigans.  At the moment the software uses the Finite Volume method to simulate the incompressible Boussinesq model on rectangular domains, without a free-surface.  There is an immersed boundary method that is being developed, which should make a free-surface possible, but that is further down the road.

There is a one-layer Shallow Water model that exists and is being further developed, which does have a free-surface.  However, that is only for a homogeneous fluid and might be too simple for your needs, depending on what questions you are aiming to answer.

**Reply by glwagner** [reply]
Woww I had not considered that immersed boundaries would allows us to simulate a free surface, but it does seem possible provided we also integrate the motion of the free surface simultaneous to the dynamics... ? I think we would have to work on the pressure solver a bit, since moving immersed boundaries may present some extra difficulties.

**Reply by francispoulin** [reply]
I agree that it would be a big step, but in time I think it should be possible  I think a much easier first step would be to have a linear free-surface, which the MITgcm does.  That includes a free-surface without actually deforming the grid.  Not a trivial change though, as a new variable needs to be introduced.  But considering that we have generalized the model structure, it should be easier now than it was a few months ago.

**Reply by kishore-nori** [reply]
@francispoulin Thank you very much for your reply and the possibilities :) I am actually interested in simulating the wave breaking for the Free surface Incompressible Newtonian NSE with homogeneous (no density or temperature variation) fluid to start with. Starting from a simple wave (not much splashing) IC with periodic or reflective BCs, so I wish to have the free surface resolved, which is not possible with the Shallow water model. But I ll checkout the Shallow water model provided here for a different use case. 

The immersed boundary method seems very interesting. Does it work on its own or should it be combined with a VOF or a Level-set method to capture the complete wave folding?

**Comment by francispoulin** [comment]
The Shallow Water model does have a free-surfce and it can be well-resolved.  What it doesn't have is any variation in the vertical. The motion consists of columnar motion where the fluid can be stretched or compacted, but the horizontal velocity does not very in the vertical.   There is not much in the way of documentation yet in Oceananigans but I hope to develop that soon.  For the moment I might suggest any textbook on Geophysical Fluid Dynamics, like for example the one by Geoff Vallis.  

I am not involved in the immersed boundary method but I understand that people are working on allowing for coatlines.  So this is being used in the horizontal.  At this time, there are no plans in the near future to use this for a free-surface.

**Reply by kishore-nori** [reply]
I see.. Thank you. What I had meant regarding the Shallow water being unable to resolve free surface was only in the case where the wave has folded and the height of the water column `h(horizontal_spacial_dims...)` cannot be a set of values where there is an interface but only a single value. Thank you for the information and book suggestion :) I ll read more on this.

**Reply by francispoulin** [reply]
Agreed.  The shallow water model does not capture the breaking using traditional methods.  Furthermore, there are also the concern that you might not want to use this model since the aspect ratio for wavebreaking is no longer small.  Using a nonhydrostatic model should yield more accurate results.  I believe the MITgcm can do this, and I suspect that this has been done, but I don't know of any references.

Good luck with your research!

**Reply by kishore-nori** [reply]
Very true! The regime is outside the modelling assumptions of Shallow water approximation. 

Nice! I ll check MITgcm. It has also been done in OpenFOAM with the VOF method but I am not well versed in OpenFOAM and was looking for something in Julia ecosystem for easier development. 

Thank you very much :)

**Reply by jklymak** [reply]
You can't do breaking surface waves in the mitgcm so far as I know.  You can of course do breaking internal waves, but the fluids are not immiscible.  But you may get "close enough" making an internal wave with two layers of very different density.

**Reply by francispoulin** [reply]
Sorry I mispoke and thanks for clarifying @jklymak

**Reply by kishore-nori** [reply]
@jklymak Oh interesting and a good idea, thanks a lot! Are there any working bounds on the ratios of the densities of the two layers? In my case, Water to Air is about 1000.

**Reply by jklymak** [reply]
I don't know why there would be, other than the usual CFL criteria meaning you will need very small time steps, but that would be the same for a free-surface solver.  But I've never tried before.

**Reply by kishore-nori** [reply]
Yeah makes sense. Only not being Immiscible can be a problem. Thank you :)

**Comment by glwagner** [comment]
@kishore-nori you might check out http://basilisk.fr/ for an alternative VOF model to OpenFoam.

**Reply by kishore-nori** [reply]
@glwagner Thank you very much for suggesting this :) It feels more comprehensible and easier to learn than OpenFOAM. I ll try it out!

**Reply by glwagner** [reply]
Let us know if you decide to write a new VOF model in julia, I think it'd be very useful! 😏 😄

**Reply by kishore-nori** [reply]
That's a good idea :) I am not an expert in VOF and have never implemented it myself, but I try understanding the source code of Oceananigans especially the FV method to see where it could fit in. Are there developer tutorials for Oceananigans?

**Reply by glwagner** [reply]
We don't have developer tutorials, but that's a neat idea!
